Fix translation of atomic functions

schuessf opened the pull request:
PR #657 added support for atomic functions. This had however two problems (in the current version):
* To improve our performance, this PR tried to avoid unecessary heap variables, i.e., we tried to translate `__atomic_store_n(&x, 1, 5)` to `x := 1` inside an atomic block rather than `write~(x, 1)` (see d08b25f and in the [PR](https://github.com/ultimate-pa/ultimate/pull/657#issuecomment-1883310446)). This way the arguments were dispatched possibly multiple times, which is problematic if they have side-effects (see [atomic_load_n_sideeffects.c](https://github.com/ultimate-pa/ultimate/compare/dev...wip/fs/atomic-fix#diff-d25c8f52ac3262e385e626766bb1204bcdc757a393dfc39c4f9ae3903c818526)).
* We put everything inside an atomic-block. However, not everything is guaranteed to be executed atomically, e.g. for `__atomic_store_n(&x, expr1, expr2)` only the actual write on `x` is atomic, not the evaluation of `expr1` or `expr2`. This could have led to missed data-races (see [atomic_memory_order.c](https://github.com/ultimate-pa/ultimate/compare/dev...wip/fs/atomic-fix#diff-6ea6a44548a6065be56965bfb6edc8167a6833a1a439bc060a9694b709734be7))
* We considered the read and write to be atomic for `__atomic_load`, but only the read should be atomic (fixed in f2e474d)

~~To fix these issues, the PR adds an interface `IPointerReadWrite` (and two implementations) that allows to get the dispatched argument, a read and a write separately (s.t. we can "control" the atomicity) and that guarantees that the dispatch is only done once. This interface is the used in `StandardFunctionHandler` for the translation of atomic functions.~~

To fix these issues, the PR adds three separate method to `ExpressionResultTransformer`:
* `dispatchPointerLValue` to dispatch the pointer expression, but tries to produce `LocalLValue`
* `makePointerAssignment` writes to the given address, produces simple assignment for `LocalLValue` and write-call for `HeapLValue`
* `readPointerValue` reads from the given address, always returns an aux-var as `RValue`, assigns the `LocalLValue` to the aux-var or reads to the aux-var from the memory for `HeapLValue`

schuessf committed:
Avoid atomic and duplicate side-effects

Previously we translated C-expressions possibly twice due to the optimization for pointer reads/writes and made everything atomic (to handle stdatomic functions).
Now, we make sure that the expression is only translated once and we separate the ExpressionResult of the original expressions and the corresponding read/write, s.t. we can control the atomicity for stdatomic functions properly.
Therefore we replaced the methods by an interface IPointerReadWrite (and matching implementations).

schuessf committed:
Add tests that failed previously

schuessf committed:
Do not evaluate the memory order atomically

schuessf committed:
Add test

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/result/ExpressionResultTransformer.java:
```diff
@@ -1147,4 +1076,104 @@ ExpressionResult apply(final ExpressionResultTransformer ert, final ExpressionRe
 				final CType targetCType, final ILocation loc, final IASTNode hook);
 	}
 
+	public interface IPointerReadWrite {
```
Please document this interface and its methods. It is not clear to me what each of them is supposed to do.

maul-esel replied:
Also, just for my understanding: Could you explain how the two classes implementing this interface differ from `HeapLValue` vs `LocalLValue` ? Do we need both distinctions?

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/result/ExpressionResultTransformer.java:
```diff
@@ -1147,4 +1076,104 @@ ExpressionResult apply(final ExpressionResultTransformer ert, final ExpressionRe
+		}
+	}
+
+	private class NormalPointerReadWrite implements IPointerReadWrite {
```
Can we rename this? I think it is not clear what "normal" means here (at least, I don't know). Also, even after renaming, please document the class.

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/result/ExpressionResultTransformer.java:
```diff
@@ -1147,4 +1076,104 @@ ExpressionResult apply(final ExpressionResultTransformer ert, final ExpressionRe
+		}
+	}
+
+	private class AddressofReadWrite implements IPointerReadWrite {
```
(nitpick) `AddressOfReadWrite`

maul-esel replied:
Also, please document the class.

maul-esel reviewed (approved):
Thanks for making sure these functions work correctly!

I took a look, and it seems a good change. I was not able to check every atomic function in precise detail though. I left some comments on the general mechanism, mostly asking for documentation / explanation.

schuessf committed:
Refactor pointer dispatches: Use LocalLValue and HeapLValue instead of IPointerReadWrite

schuessf committed:
Make only the read atomic for __atomic_load

schuessf commented:
> Also, just for my understanding: Could you explain how the two classes implementing this interface differ from `HeapLValue` vs `LocalLValue` ? Do we need both distinctions?

@maul-esel This interface was not really necessary,  `HeapLValue` and `LocalLValue`  are sufficient for distinction.
Therefore I replaced this interface with separate methods and also added some documentation.
So please review this PR again.

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/result/ExpressionResultTransformer.java:
```diff
@@ -1041,110 +1039,112 @@ public ExpressionResult convertNullPointerConstantToPointer(final ExpressionResu
 		}
+		final ExpressionResultBuilder builder = new ExpressionResultBuilder(result);
+		builder.resetLrValue(LRValueFactory.constructHeapLValue(mTypeHandler, result.getLrValue().getValue(),
+				result.getCType(), null));
```
Out of interest: In which cases could it happen that we do not have a `HeapLValue` here, and what do we have instead?

schuessf replied:
This happens for example if the expression is of the form `&x`, where `x` is already on the heap (we get an `RValue` in that case) or if the expression is some variable `p` of pointer type (we get a `LocalLValue` in that case). Actually, I did not find a case, where we get a `HeapLValue` here, but I will keep that case anyway (`result.getLrValue().getValue()` would throw an exception, if this happens).
